pkg/witness/publish_checkpoint.go

```diff
@@ -0,0 +1,196 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package witness
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/google/trillian"
+	"github.com/google/trillian/types"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/redis/go-redis/v9"
+	"github.com/sigstore/rekor/pkg/log"
+	"github.com/sigstore/rekor/pkg/trillianclient"
+	"github.com/sigstore/rekor/pkg/util"
+	"github.com/sigstore/sigstore/pkg/signature"
+	"google.golang.org/grpc/codes"
+)
+
+// CheckpointPublisher is a long-running job to periodically publish signed checkpoints to etc.d
+type CheckpointPublisher struct {
+	ctx context.Context
+	// logClient is the client for Trillian
+	logClient trillian.TrillianLogClient
+	// treeID is used to construct the origin and configure the Trillian client
+	treeID int64
+	// hostname is used to construct the origin ("hostname - treeID")
+	hostname string
+	// signer signs the checkpoint
+	signer signature.Signer
+	// publishFreq is how often a new checkpoint is published to Rekor, in minutes
+	checkpointFreq uint
+	// redisClient to upload signed checkpoints
+	redisClient *redis.Client
+	// reqCounter tracks successes and failures for publishing
+	reqCounter *prometheus.CounterVec
+}
+
+// Constant values used with metrics
+const (
+	Success = iota
+	SuccessObtainLock
+	GetCheckpoint
+	UnmarshalCheckpoint
+	SignCheckpoint
+	RedisFailure
+	RedisLatestFailure
+)
+
+// NewCheckpointPublisher creates a CheckpointPublisher to write stable checkpoints to Redis
+func NewCheckpointPublisher(ctx context.Context,
+	logClient trillian.TrillianLogClient,
+	treeID int64,
+	hostname string,
+	signer signature.Signer,
+	redisClient *redis.Client,
+	checkpointFreq uint,
+	reqCounter *prometheus.CounterVec) CheckpointPublisher {
+	return CheckpointPublisher{ctx: ctx, logClient: logClient, treeID: treeID, hostname: hostname,
+		signer: signer, checkpointFreq: checkpointFreq, redisClient: redisClient, reqCounter: reqCounter}
+}
+
+// StartPublisher creates a long-running task that publishes the latest checkpoint every X minutes
+// Writing to Redis is best effort. Failure will be detected either through metrics or by witnesses
+// or Verifiers monitoring for fresh checkpoints. Failure can occur after a lock is obtained but
+// before publishing the latest checkpoint. If this occurs due to a sporadic failure, this simply
+// means that a witness will not see a fresh checkpoint for an additional period.
+func (c *CheckpointPublisher) StartPublisher() {
+	tc := trillianclient.NewTrillianClient(context.Background(), c.logClient, c.treeID)
+	sTreeID := strconv.FormatInt(c.treeID, 10)
+
+	// publish on startup to ensure a checkpoint is available the first time Rekor starts up
+	c.publish(&tc, sTreeID)
+
+	ticker := time.NewTicker(time.Duration(c.checkpointFreq) * time.Minute)
+	go func() {
```

Comment:
you're leaking this goroutine as there's no way to quit it. 

You probably want to fix this, and also add the following into your unit test file to ensure it doesn't regress:
```go
import (
...
        "go.uber.org/goleak"
)

func TestMain(m *testing.M) {
        goleak.VerifyTestMain(m)
}
```

Reply:
Fixed by adding a context and calling cancel on shutdown